### ISSUE TITLE
glyph.export is not used when writing ufos

### ISSUE THREAD

**Issue by schriftgestalt** [googlei18n/glyphsLib#320]
When glyph.export is False, the glyph should not be written to the final font file. 

And, if a component points to it, it this component needs to be decomposed.

**Comment by belluzj** [comment]
I think the current strategy here in `glyphsLib` is to write all glyphs to UFO and write as well their "export" flag. The export flag is then used by `fontmake` to filter the glyphs in the built font:

https://github.com/googlei18n/fontmake/blob/master/Lib/fontmake/font_project.py#L301

Are you proposing that `glyphsLib` should directly remove the glyphs from the UFOs?

**Comment by schriftgestalt** [comment]
I don’t really care who is doing it. Right now, nobody is taking care of it. And I have implemented in glyphsLib almost.

**Comment by schriftgestalt** [comment]
I was using fontmake to generate a variable font from a .glyphs file. It failed because there are some glyphs that have incompatible outlines. Those are set to not export. 

After only putting the right glyphs in the ufo, it still complains about the kerning classes as I have not yet fixed that. It is much easier to deal with the kerning classes in glyphsLib instead of cleaning it up in fontmake.

**Comment by moyogo** [comment]
Aren't the glyphs that are not exported also muted in the designspace?

**Comment by belluzj** [comment]
> I’m trying to run the builder directly in Glyphs

Cool! Did you run the one from the master branch or the one from the `real-ufo-roundtrip` pull request?

**Comment by schriftgestalt** [comment]
I stuffed all my changes in this repo. I’ll make nice commits when I’m done. https://github.com/schriftgestalt/glyphsLib/tree/some-adjustments

**Comment by schriftgestalt** [comment]
I see that is a `load_to_ufos` function. How would a `load_to_glyphs` look like that takes an path to a designspace file as argument?

**Comment by belluzj** [comment]
It's in this PR: https://github.com/googlei18n/glyphsLib/pull/244

Have a look at the README from my branch which has examples of how to go from designspace to Glyphs: https://github.com/belluzj/glyphsLib/blob/real-ufo-roundtrip/README.rst#go-back-and-forth-between-ufos-and-glyphs

**Comment by anthrotype** [comment]
> the current strategy here in glyphsLib is to write all glyphs to UFO and write as well their "export" flag. The export flag is then used by fontmake to filter the glyphs in the built font

Right. And fontmake is using the fonttools subsetter only _after_ ufo2ft has generated the full TTFs that also include the non-export glyphs. I know it sounds convoluted, but it allows us to rely on the subsetter to decompose the components whose base glyph was dropped, and subset the GSUB/GPOS table (without having to do that on the features.fea ourselves).

The conversion to quadratic is done in cu2qu and the latter doesn't know about these export flags, simply tries to convert glyphs with same name. Maybe it should be made aware of them, and skip.

**Comment by belluzj** [comment]
@schriftgestalt  I have introduced three functions that I use extensively in test:

```typescript
to_designspace(GSFont) -> DesignspaceDocument
to_ufos(GSFont) -> List<defcon.Font>
to_glyphs(DesignspaceDocument | List<defcon.Font>) -> GSFont
```

**Comment by anthrotype** [comment]
As I'm gradually moving more and more of the functionalities from fontmake (the compiler CLI) to ufo2ft (the compiler library), I think this subsetting-by-export-flag step should also move to ufo2ft.postProcessor

**Comment by schriftgestalt** [comment]
> I think this subsetting-by-export-flag step should also move to ufo2ft.postProcessor

as long as someone takes care of it..
